File: gui/src/api/onboarding.ts
/**
 * API client for onboarding flow
 *
 * v1 workspace-scoped routes for multi-tenant strategy builder.
 * Legacy user-scoped routes are deprecated and will be removed.
 */

import { apiDelete, apiGet, apiPost } from './client'
import { getCurrentWorkspaceId } from '../stores/workspaceStore'

// =============================================================================
// Types
// =============================================================================

export interface OnboardingQuestion {
  id: string
  question: string
  placeholder?: string
  type: 'text' | 'select'
  options?: Array<{
    id: string
    name: string
    description: string
  }>
}

export interface StartResponse {
  user_id: string
  questions: OnboardingQuestion[]
}

export interface GeneratedPost {
  post_number: number
  topic: string
  shape: string
  cadence: string
}

export interface GeneratedChapter {
  chapter_number: number
  title: string
  theme: string
  theme_description: string
  posts: GeneratedPost[]
}

export interface GeneratedPlan {
  signature_thesis: string
  chapters: GeneratedChapter[]
}

export interface QuickModeResponse {
  plan: GeneratedPlan
}

export interface DeepModeMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface DeepModeState {
  messages: DeepModeMessage[]
  turn_count: number
  ready_to_generate: boolean
}

export interface DeepModeResponse {
  assistant_message: string
  state: DeepModeState
  ready_to_generate: boolean
}

export interface ApproveResponse {
  goal_id: string
  chapter_count: number
  post_count: number
}

// =============================================================================
// Start Onboarding
// =============================================================================

export async function startOnboarding(
  name: string,
  email?: string
): Promise<StartResponse> {
  return apiPost<StartResponse>('/onboarding/start', { name, email })
}

export async function fetchQuickQuestions(): Promise<{ questions: OnboardingQuestion[] }> {
  return apiGet<{ questions: OnboardingQuestion[] }>('/onboarding/questions')
}

// =============================================================================
// Quick Mode
// =============================================================================

export interface QuickModeAnswers {
  user_id: string
  professional_role: string
  known_for: string
  target_audience: string
  content_style: string
  posts_per_week: number
}

export async function submitQuickMode(
  answers: QuickModeAnswers
): Promise<QuickModeResponse> {
  return apiPost<QuickModeResponse>('/onboarding/quick', answers)
}

// =============================================================================
// Deep Mode
// =============================================================================

export async function startDeepDiscovery(): Promise<DeepModeResponse> {
  return apiPost<DeepModeResponse>('/onboarding/deep/start')
}

export async function sendDeepMessage(
  userId: string,
  message: string,
  state: DeepModeState,
  forceReady?: boolean
): Promise<DeepModeResponse> {
  return apiPost<DeepModeResponse>('/onboarding/deep/message', {
    user_id: userId,
    message,
    state,
    force_ready: Boolean(forceReady),
  })
}

export async function generateFromDeep(
  userId: string,
  contentStyle: string,
  state: DeepModeState
): Promise<QuickModeResponse> {
  return apiPost<QuickModeResponse>('/onboarding/deep/generate', {
    user_id: userId,
    content_style: contentStyle,
    state,
  })
}

// =============================================================================
// Plan Approval
// =============================================================================

export interface ApprovePlanRequest {
  user_id: string
  plan: GeneratedPlan
  positioning: string
  target_audience: string
  content_style: string
  onboarding_mode: 'quick' | 'deep' | 'chat'
  transcript?: string
  workspace_id?: string | null
}

export async function approvePlan(
  request: ApprovePlanRequest
): Promise<ApproveResponse> {
  return apiPost<ApproveResponse>('/onboarding/approve', request)
}

// =============================================================================
// Existing Strategy
// =============================================================================

export interface ExistingChapter {
  id: string
  chapter_number: number
  title: string
  description: string | null
  theme: string | null
  theme_description: string | null
  weeks_start: number | null
  weeks_end: number | null
  post_count: number
  completed_count: number
}

export interface ExistingGoal {
  id: string
  workspace_id: string
  strategy_type: 'series' | 'daily' | 'campaign'
  positioning: string | null
  signature_thesis: string | null
  target_audience: string | null
  content_style: string | null
  voice_profile_id: string | null
  image_config_set_id: string | null
}

export interface StrategySummary {
  total_chapters: number
  total_posts: number
  completed_posts: number
  weeks_total: number
}

export interface ExistingStrategy {
  exists: boolean
  goal?: ExistingGoal
  chapters?: ExistingChapter[]
  summary?: StrategySummary
}

export async function fetchExistingStrategy(): Promise<ExistingStrategy> {
  const workspaceId = getCurrentWorkspaceId()
  if (!workspaceId) {
    return { exists: false }
  }

  try {
    // Fetch goals and chapters from workspace-scoped endpoints
    const [goals, chapters] = await Promise.all([
      apiGet<ExistingGoal[]>(`/v1/w/${workspaceId}/goals`),
      apiGet<ExistingChapter[]>(`/v1/w/${workspaceId}/chapters`),
    ])

    if (!goals.length) {
      return { exists: false }
    }

    // Use the first goal as the primary strategy
    const goal = goals[0]

    // Calculate summary
    const totalPosts = chapters.reduce((sum, ch) => sum + ch.post_count, 0)
    const completedPosts = chapters.reduce((sum, ch) => sum + ch.completed_count, 0)
    const weeksTotal = chapters.reduce((max, ch) => Math.max(max, ch.weeks_end || 0), 0)

    return {
      exists: true,
      goal,
      chapters,
      summary: {
        total_chapters: chapters.length,
        total_posts: totalPosts,
        completed_posts: completedPosts,
        weeks_total: weeksTotal,
      },
    }
  } catch {
    return { exists: false }
  }
}

// =============================================================================
// v1 Workspace-Scoped API (preferred)
// =============================================================================

export interface V1ContentStyle {
  id: string
  name: string
  description: string
}

export interface V1QuickModeRequest {
  professional_role: string
  known_for: string
  target_audience: string
  content_style: string
  posts_per_week: number
}

export interface V1DeepModeMessageRequest {
  message: string
  state?: DeepModeState | null
  force_ready?: boolean
}

export interface V1DeepModeResponse {
  message: string
  state: DeepModeState
  is_ready: boolean
}

export interface V1GeneratePlanRequest {
  content_style: string
  state: DeepModeState
}

export interface V1ApprovePlanRequest {
  plan: GeneratedPlan
  positioning: string
  target_audience: string
  content_style: string
  onboarding_mode: 'quick' | 'deep' | 'chat'
  transcript?: string | null
}

export interface V1ApproveResponse {
  goal_id: string
  chapters: string[]
}

export interface V1StrategyResponse {
  goal: ExistingGoal | null
  chapters: ExistingChapter[]
}

/**
 * Get available content styles (workspace-scoped).
 */
export async function fetchContentStylesV1(
  workspaceId: string
): Promise<{ styles: V1ContentStyle[] }> {
  return apiGet<{ styles: V1ContentStyle[] }>(
    `/v1/w/${workspaceId}/onboarding/content-styles`
  )
}

/**
 * Submit quick mode onboarding (workspace-scoped).
 */
export async function submitQuickModeV1(
  workspaceId: string,
  request: V1QuickModeRequest
): Promise<QuickModeResponse> {
  return apiPost<QuickModeResponse>(
    `/v1/w/${workspaceId}/onboarding/quick`,
    request
  )
}

/**
 * Send deep mode message (workspace-scoped).
 */
export async function sendDeepMessageV1(
  workspaceId: string,
  request: V1DeepModeMessageRequest
): Promise<V1DeepModeResponse> {
  return apiPost<V1DeepModeResponse>(
    `/v1/w/${workspaceId}/onboarding/deep/message`,
    request
  )
}

/**
 * Generate plan from deep mode (workspace-scoped).
 */
export async function generateFromDeepV1(
  workspaceId: string,
  request: V1GeneratePlanRequest
): Promise<QuickModeResponse> {
  return apiPost<QuickModeResponse>(
    `/v1/w/${workspaceId}/onboarding/deep/generate-plan`,
    request
  )
}

/**
 * Approve and save generated plan (workspace-scoped).
 */
export async function approvePlanV1(
  workspaceId: string,
  request: V1ApprovePlanRequest
): Promise<V1ApproveResponse> {
  return apiPost<V1ApproveResponse>(
    `/v1/w/${workspaceId}/onboarding/approve`,
    request
  )
}

/**
 * Get current strategy for workspace.
 */
export async function fetchStrategyV1(
  workspaceId: string
): Promise<V1StrategyResponse> {
  return apiGet<V1StrategyResponse>(
    `/v1/w/${workspaceId}/onboarding/strategy`
  )
}

/**
 * Delete current strategy for workspace.
 */
export async function deleteStrategyV1(workspaceId: string): Promise<void> {
  return apiDelete(`/v1/w/${workspaceId}/onboarding/strategy`)
}

File: api/routes/deprecated/onboarding.py
"""API routes for onboarding flow."""

from typing import Annotated, Optional
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from api.auth.dependencies import CurrentUser, get_current_user
from api.services.content_service import ContentService
from api.services.onboarding_service import (
    OnboardingService,
    QuickOnboardingAnswers,
    DeepModeState,
    GeneratedPlan,
)
from runner.db.models import UserRole

router = APIRouter(prefix="/onboarding", tags=["onboarding"])
onboarding_service = OnboardingService()
content_service = ContentService()


def _verify_user_access(current_user: CurrentUser, target_user_id: UUID) -> None:
    """Verify user can access the target user's data."""
    is_owner = current_user.user.role == UserRole.owner
    if str(current_user.user_id) != str(target_user_id) and not is_owner:
        raise HTTPException(status_code=404, detail="Resource not found")


# =============================================================================
# Request Models
# =============================================================================


class StartRequest(BaseModel):
    """Start onboarding - creates user."""

    name: str
    email: Optional[str] = None


class QuickModeRequest(BaseModel):
    """Quick mode answers."""

    user_id: UUID
    professional_role: str
    known_for: str
    target_audience: str
    content_style: str
    posts_per_week: int


class DeepModeMessageRequest(BaseModel):
    """User message in deep mode conversation."""

    user_id: UUID
    message: str
    state: Optional[DeepModeState] = None  # Pass state for continuation
    force_ready: bool = False  # User can manually trigger generation readiness


class GeneratePlanRequest(BaseModel):
    """Request to generate plan from deep mode."""

    user_id: UUID
    content_style: str
    state: DeepModeState


class ApprovePlanRequest(BaseModel):
    """Approve and save generated plan."""

    user_id: UUID
    plan: dict  # GeneratedPlan as dict
    positioning: str
    target_audience: str
    content_style: str
    onboarding_mode: str
    transcript: Optional[str] = None
    workspace_id: Optional[UUID] = None


# =============================================================================
# Response Models
# =============================================================================


class StartResponse(BaseModel):
    """Response from starting onboarding."""

    user_id: str
    questions: list[dict]


class QuickModeResponse(BaseModel):
    """Response from quick mode - generated plan."""

    plan: GeneratedPlan


class DeepModeResponse(BaseModel):
    """Response from deep mode - assistant message + state."""

    assistant_message: str
    state: DeepModeState
    ready_to_generate: bool


class ApproveResponse(BaseModel):
    """Response from plan approval."""

    goal_id: str
    chapter_count: int
    post_count: int


# =============================================================================
# Endpoints
# =============================================================================


@router.get("/strategy/{user_id}")
def get_existing_strategy(
    user_id: UUID,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
    platform_id: Optional[UUID] = None,
):
    """Get existing strategy for a user, optionally filtered by platform."""
    _verify_user_access(current_user, user_id)

    goal = content_service.get_goal(user_id, platform_id=platform_id)
    if not goal:
        return {"exists": False}

    chapters = content_service.get_chapters(user_id, platform_id=platform_id)
    total_posts = sum(c.post_count for c in chapters)
    total_completed = sum(c.completed_count for c in chapters)

    return {
        "exists": True,
        "goal": {
            "id": goal.id,
            "strategy_type": goal.strategy_type,
            "positioning": goal.positioning,
            "signature_thesis": goal.signature_thesis,
            "target_audience": goal.target_audience,
            "content_style": goal.content_style,
            "onboarding_mode": goal.onboarding_mode,
        },
        "summary": {
            "total_chapters": len(chapters),
            "total_posts": total_posts,
            "completed_posts": total_completed,
            "weeks_total": chapters[-1].weeks_end if chapters else 0,
        },
        "chapters": [
            {
                "id": c.id,
                "chapter_number": c.chapter_number,
                "title": c.title,
                "description": c.description,
                "theme": c.theme,
                "theme_description": c.theme_description,
                "weeks_start": c.weeks_start,
                "weeks_end": c.weeks_end,
                "post_count": c.post_count,
                "completed_count": c.completed_count,
            }
            for c in chapters
        ],
    }


@router.post("/start", response_model=StartResponse)
def start_onboarding(
    request: StartRequest,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
):
    """Start onboarding flow - creates user and returns questions."""
    user_id = content_service.get_or_create_user(request.name, request.email)
    questions = OnboardingService.get_quick_mode_questions()
    return StartResponse(user_id=user_id, questions=questions)


@router.get("/questions")
def get_quick_questions(
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
):
    """Get quick mode questions."""
    return {"questions": OnboardingService.get_quick_mode_questions()}


# =============================================================================
# Quick Mode
# =============================================================================


@router.post("/quick", response_model=QuickModeResponse)
def quick_onboarding(
    request: QuickModeRequest,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
):
    """Process quick mode answers and generate plan."""
    _verify_user_access(current_user, request.user_id)

    try:
        answers = QuickOnboardingAnswers(
            professional_role=request.professional_role,
            known_for=request.known_for,
            target_audience=request.target_audience,
            content_style=request.content_style,
            posts_per_week=request.posts_per_week,
        )
        plan = onboarding_service.generate_quick_plan(answers)
        return QuickModeResponse(plan=plan)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to generate plan: {e}")


# =============================================================================
# Deep Mode
# =============================================================================


@router.post("/deep/start", response_model=DeepModeResponse)
def start_deep_discovery(
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
):
    """Start deep discovery conversation."""
    try:
        state = onboarding_service.start_deep_discovery()
        return DeepModeResponse(
            assistant_message=state.messages[-1].content,
            state=state,
            ready_to_generate=state.ready_to_generate,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to start discovery: {e}")


@router.post("/deep/message", response_model=DeepModeResponse)
def send_deep_message(
    request: DeepModeMessageRequest,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
):
    """Send message in deep discovery conversation.

    Set force_ready=true to manually trigger generation readiness
    (useful if LLM doesn't offer to generate after enough turns).
    """
    _verify_user_access(current_user, request.user_id)

    if not request.state:
        raise HTTPException(status_code=400, detail="State required for continuation")

    try:
        state = onboarding_service.continue_deep_discovery(
            request.state,
            request.message,
            force_ready=request.force_ready,
        )
        # If force_ready, last message is user's, not assistant's
        assistant_message = (
            state.messages[-1].content
            if state.messages[-1].role == "assistant"
            else "Ready to generate your content strategy!"
        )
        return DeepModeResponse(
            assistant_message=assistant_message,
            state=state,
            ready_to_generate=state.ready_to_generate,
        )
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to continue discovery: {e}"
        )


@router.post("/deep/generate", response_model=QuickModeResponse)
def generate_from_deep(
    request: GeneratePlanRequest,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
):
    """Generate plan from deep discovery conversation."""
    _verify_user_access(current_user, request.user_id)

    try:
        plan = onboarding_service.generate_deep_plan(
            request.state,
            request.content_style,
        )
        return QuickModeResponse(plan=plan)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to generate plan: {e}")


# =============================================================================
# Plan Approval
# =============================================================================


@router.post("/approve", response_model=ApproveResponse)
def approve_plan(
    request: ApprovePlanRequest,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
):
    """Approve and save generated plan to database."""
    _verify_user_access(current_user, request.user_id)

    try:
        plan = GeneratedPlan(**request.plan)
        goal_id = onboarding_service.save_plan(
            user_id=request.user_id,
            plan=plan,
            positioning=request.positioning,
            target_audience=request.target_audience,
            content_style=request.content_style,
            onboarding_mode=request.onboarding_mode,
            onboarding_transcript=request.transcript,
            workspace_id=request.workspace_id,
        )

        # Count chapters and posts
        chapter_count = len(plan.chapters)
        post_count = sum(len(c.get("posts", [])) for c in plan.chapters)

        return ApproveResponse(
            goal_id=goal_id,
            chapter_count=chapter_count,
            post_count=post_count,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to save plan: {e}")
